cdap-authorization-ldap-role/src/main/java/io/cdap/cdap/security/authorization/ldap/role/searcher/LDAPSearcher.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright © 2021-2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.security.authorization.ldap.role.searcher;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Hashtable;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.naming.Context;
+import javax.naming.NamingEnumeration;
+import javax.naming.NamingException;
+import javax.naming.PartialResultException;
+import javax.naming.directory.Attribute;
+import javax.naming.directory.DirContext;
+import javax.naming.directory.InitialDirContext;
+import javax.naming.directory.SearchControls;
+import javax.naming.directory.SearchResult;
+
+import static javax.naming.directory.SearchControls.SUBTREE_SCOPE;
+
+
+/**
+ * Service for searching user's groups in LDAP
+ */
+public class LDAPSearcher {
+  private static final Logger LOG = LoggerFactory.getLogger(LDAPSearcher.class);
+
+  private final Hashtable<String, String> properties;
+  private final LDAPSearchConfig config;
+  private final String[] baseDNList;
+
+  /**
+   * Constructor
+   *
+   * @param config {@link LDAPSearchConfig} configuration for LDAP searcher
+   */
+  public LDAPSearcher(LDAPSearchConfig config) {
+    this.config = config;
+    properties = getConnectionProperties();
+    baseDNList = config.getSearchBaseDn().split(LDAPConstants.BASE_DN_SPLITTER);
+  }
+
+  /**
+   * Tests connection to LDAP
+   */
+  public void testConnection() {
+    try {
+      DirContext context = getConnection();
+      context.close();
+    } catch (NamingException e) {
+      String errorMsg = String.format("Failed to establish connection to '%s'", config.getUrl());
+      throw new RuntimeException(errorMsg, e);
+    }
+  }
+
+  /**
+   * Searches for groups by username
+   *
+   * @param username Name of user
+   * @return Set of groups
+   */
+  public Set<String> searchGroups(String username) {
```

Comment:
The searchGroups functionality seems like it is important enough to warrant a unit test. Can we add an interface behind the `LDAPSearcher` which abstracts out connecting to the LDAP, perhaps called `LDAPClient`? Then, we can add some tests for the `LDAPSearcher` functionality. Thanks!

Reply:
Added interface `LDAPClient`, it's implementation and tests for `LDAPSearcher` with mocks of `LDAPClient`